workbench/src/wb_project.c

```diff
@@ -0,0 +1,1491 @@
+		{
+			gchar **bookmarks_strings, *file, *rel_path;
+
+			bookmarks_strings = g_new0(gchar *, boomarks_size+1);
+			for (index = 0 ; index < boomarks_size ; index++ )
+			{
+				file = wb_project_get_bookmark_at_index(prj, index);
+				rel_path = get_any_relative_path(prj->filename, file);
+
+				bookmarks_strings[index] = rel_path;
+			}
+			g_key_file_set_string_list
+				(kf, "Workbench", "Bookmarks", (const gchar **)bookmarks_strings, boomarks_size);
+			for (index = 0 ; index < boomarks_size ; index++ )
+			{
+				g_free (bookmarks_strings[index]);
+			}
+			g_free(bookmarks_strings);
+		}
+
+		/* Init tmp data */
+		tmp.kf = kf;
+		tmp.dir_count = 1;
+
+		/* Store our directories */
+		g_slist_foreach(prj->directories, (GFunc)wb_project_save_directories, &tmp);
+
+		/* Get data as string */
+		contents = g_key_file_to_data (kf, &length, error);
+		g_key_file_free(kf);
+
+		/* Save to file */
+		success = g_file_set_contents (prj->filename, contents, length, error);
+		if (success)
+		{
+			prj->modified = FALSE;
+		}
+		g_free (contents);
+	}
+	else if (error != NULL)
+	{
+		g_set_error (error, 0, 0,
+					 "Internal error: param missing (file: %s, line %d)",
+					 __FILE__, __LINE__);
+	}
+
+	return success;
+}
+
+
+/** Load a project.
+ *
+ * The function loads the project data from file @a filename into the
+ * project structure @a prj.
+ *
+ * @param prj      The project
+ * @param filename File to load
+ * @param error    Location to store error info at
+ * @return TRUE on success, FALSE otherwise
+ *
+ **/
+gboolean wb_project_load(WB_PROJECT *prj, gchar *filename, GError **error)
+{
+	gboolean success = FALSE;
+
+	if (prj != NULL)
+	{
+		GKeyFile *kf;
+		guint	 index;
+		gchar    *contents;
+		gchar    key[100];
+		gsize    length;
+
+		if (!g_file_get_contents (filename, &contents, &length, error))
+		{
+			return FALSE;
+		}
+
+		kf = g_key_file_new ();
+
+		if (!g_key_file_load_from_data (kf, contents, length,
+					G_KEY_FILE_KEEP_COMMENTS | G_KEY_FILE_KEEP_TRANSLATIONS,
+					error))
+		{
+			g_key_file_free (kf);
+			g_free (contents);
+			return FALSE;
+		}
+
+		if (g_key_file_has_group (kf, "Workbench"))
+		{
+			WB_PROJECT_DIR *new_dir;
+			gchar *str;
+			gchar **splitv, **bookmarks_strings;
+
+			/* Load project bookmarks from string list */
+			bookmarks_strings = g_key_file_get_string_list (kf, "Workbench", "Bookmarks", NULL, error);
+			if (bookmarks_strings != NULL)
+			{
+				gchar **file, *abs_path;
+
+				file = bookmarks_strings;
+				while (*file != NULL)
+				{
+					abs_path = get_combined_path(prj->filename, *file);
+					if (abs_path != NULL)
+					{
+						wb_project_add_bookmark_int(prj, abs_path);
+						g_free(abs_path);
+					}
+					file++;
+				}
+				g_strfreev(bookmarks_strings);
+			}
+
+			/* Load project dirs */
+			for (index = 1 ; index < 1025 ; index++)
+			{
+				g_snprintf(key, sizeof(key), "Dir%u-BaseDir", index);
+				if (!g_key_file_has_key (kf, "Workbench", key, NULL))
```

Comment:
this isn't really useful as [`g_key_file_get_string()`](https://developer.gnome.org/glib/stable/glib-Key-value-file-parser.html#g-key-file-get-string) (below) returns `NULL` if the key is not found